crates/core/src/factories/core_bootstrap.rs

```diff
@@ -0,0 +1,296 @@
+    server_url: Arc<str>,
+    push_send: PushSend,
+    mut push_recv: PushRecv,
+) {
+    let mut wait = None;
+
+    while let Some(info) = push_recv.recv().await {
+        let url =
+            format!("{server_url}/bootstrap/{}/{}", &info.space, &info.agent);
+        let enc = match info.encode() {
+            Err(_) => continue,
+            Ok(enc) => enc,
+        };
+        match tokio::task::spawn_blocking(move || {
+            ureq::put(&url).send_string(&enc)
+        })
+        .await
+        {
+            Ok(Ok(_)) => {
+                // the put was successful, we don't need to wait
+                // before sending the next info if it is ready
+                wait = None;
+            }
+            _ => {
+                let now = Timestamp::now();
+
+                // the put failed, send it back to try again if not expired
+                if info.expires_at > now {
+                    let _ = push_send.try_send(info);
+                }
+
+                // we need to configure a backoff so we don't hammer the server
+                match wait {
+                    None => wait = Some(config.backoff_min()),
+                    Some(p) => {
+                        let mut p = p * 2;
+                        if p > config.backoff_max() {
+                            p = config.backoff_max();
+                        }
+                        wait = Some(p);
+                    }
+                }
+
+                // wait for the backoff time
+                if let Some(wait) = &wait {
+                    tokio::time::sleep(*wait).await;
+                }
+            }
+        }
+    }
+}
+
+async fn poll_task(
+    builder: Arc<builder::Builder>,
+    config: CoreBootstrapConfig,
+    server_url: Arc<str>,
+    space: SpaceId,
+    peer_store: peer_store::DynPeerStore,
+) {
+    let mut wait = config.backoff_min();
+
+    loop {
+        let url = format!("{server_url}/bootstrap/{space}");
+        if let Ok(Ok(data)) = tokio::task::spawn_blocking(move || {
+            ureq::get(&url)
+                .call()
+                .map_err(std::io::Error::other)?
+                .into_string()
+        })
+        .await
+        {
+            if let Ok(list) = agent::AgentInfoSigned::decode_list(
+                &builder.verifier,
+                data.as_bytes(),
+            ) {
+                // count decoding a success, and set the wait to max
+                wait = config.backoff_max();
+
+                let list = list
+                    .into_iter()
+                    .filter_map(|l| match l {
+                        Ok(l) => Some(l),
+                        Err(_) => None,
+                    })
+                    .collect::<Vec<_>>();
+
+                let _ = peer_store.insert(list).await;
+            }
+        }
+
+        wait *= 2;
+        if wait > config.backoff_max() {
+            wait = config.backoff_max();
+        }
+
+        tokio::time::sleep(wait).await;
```

Comment:
Should `wait` be reset to `config.backoff_min()` after the sleep? Otherwise it gets set to max on the first success and stays there forever

Reply:
The other push task loop does that. I feel like this one is less dire... once we're already up and running (i.e. we've had a single bootstrap success), we should be gossiping agents and not rely on bootstrapping anymore. I was thinking even if we missed a poll once in five minutes, it'd be fine to just try again five minutes later.

Happy to change this if you think it is needed.

Reply:
Makes sense to me. Errors 5 minutes apart seems relatively unlikely unless something is also going to be wrong during that interval too